Reject unsupported, unstable ABIs during AST lowering

workingjubilee opened the pull request:
Currently, we check whether ABIs have support on the target we are compiling for during HIR analysis. This unfortunately is a slightly "whack-a-mole" proposition, because there is no inherent relationship between ABI strings and their usages in the HIR[^1]. We can instead check for support in `rustc_ast_lowering`, where we first materialize `ExternAbi` from strings. This lets us always error on ABIs we wish to error on. 

As a first step, we throw the switch for unstable ABIs. Despite being nightly features, we actually have emitted future compatibility warnings for all unsupported ABIs for some time:
- `unsupported_calling_conventions` became a hard error.
- `unsupported_fn_ptr_calling_conventions` is warn-in-deps since Rust 1.87, a stable Rust version.

With this we effectively make the latter lint a hard error on *unstable* ABIs. Stable ABIs still only receive warnings.

As a bonus, this addresses three ICEs that the compiler could reach. It prevents many more LLVM errors from being seen by programmers as well, as most unstable ABIs cannot have correct code generated for them by targets that do not support the ABI.

Fixes rust-lang/rust#132430
Fixes rust-lang/rust#138738

[^1]: For example, we already missed some use-sites of ABI strings because FnSigs are not BareFnTys, requiring us to implement a new future-compat-warning to handle the function pointer case.

workingjubilee committed:
compiler: plug unsupported ABI leakage from the AST

We elaborate rustc_ast_lowering to prevent unstable, unsupported ABIs
from leaking through the HIR without being checked for target support.
Previously ad-hoc checking on various HIR items required making sure
we check every HIR item which could contain an extern "{abi}" string.
This is a losing proposition compared to gating the lowering itself.

As a consequence, unstable ABI strings will now hard-error instead of
triggering the FCW `unsupported_fn_ptr_calling_conventions`.
This FCW was upgraded to warn in dependencies in Rust 1.87 so it has
become active within a stable Rust version and it was within the
compiler's nightly-feature contract to break this code without warning,
so this breakage has likely had a reasonable amount of foreshadowing.

workingjubilee committed:
compiler: limit in-HIR ABI support checks to stable ABIs

We already gated unstable, unsupported ABIs during AST lowering.
If we check all ABIs for target support during HIR lowering, then we
risk duplicate errors in code that uses unstable extern "{abi}"s.

Limiting the ABI support checks to stable ABIs reduces this, and
remains correct because of gating all unstable ABIs already.
This code complication can be reduced when we switch the FCW to error.

workingjubilee committed:
tests: Rework abi_gpu_kernel feature gate test

Explicitly test it for relevant targets and check it errors on hosts.

workingjubilee committed:
tests: Upgrade unstable ABIs to errors in unsupported ABI test

workingjubilee committed:
tests: Update and bless cmse-nonsecure ABI tests

workingjubilee committed:
tests: Adopt ABI transmute tests from crashtests

workingjubilee committed:
tests: Adopt unsupported ABI in weird type impl test

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15382962114/job/43276612298) [(plain)](https://github.com/rust-lang/rust/commit/41dea87b794b559954ed903c3cea07bb17fb459a/checks/43276612298/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [rustdoc-json] tests/rustdoc-json/fn_pointer/abi.rs stdout ----

error: rustdoc failed!
status: exit status: 1
command: cd "/checkout/obj/build/aarch64-unknown-linux-gnu/test/rustdoc-json/fn_pointer/abi" && env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustdoc" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/rustdoc-json/fn_pointer/abi/auxiliary" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/rustdoc-json/fn_pointer/abi" "--deny" "warnings" "/checkout/tests/rustdoc-json/fn_pointer/abi.rs" "-A" "internal_features" "--output-format" "json" "-Zunstable-options"
stdout: none
--- stderr -------------------------------
error[E0570]: `"vectorcall"` is not a supported ABI for the current target
##[error]  --> /checkout/tests/rustdoc-json/fn_pointer/abi.rs:19:32
   |
19 | pub type AbiVecorcall = extern "vectorcall" fn();
   |                                ^^^^^^^^^^^^

error[E0570]: `"vectorcall-unwind"` is not a supported ABI for the current target
##[error]  --> /checkout/tests/rustdoc-json/fn_pointer/abi.rs:22:38
   |
22 | pub type AbiVecorcallUnwind = extern "vectorcall-unwind" fn();
   |                                      ^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0570`.
------------------------------------------


---- [rustdoc-json] tests/rustdoc-json/fns/abi.rs stdout ----

error: rustdoc failed!
status: exit status: 1
command: cd "/checkout/obj/build/aarch64-unknown-linux-gnu/test/rustdoc-json/fns/abi" && env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustdoc" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/rustdoc-json/fns/abi/auxiliary" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/rustdoc-json/fns/abi" "--deny" "warnings" "/checkout/tests/rustdoc-json/fns/abi.rs" "-A" "internal_features" "--output-format" "json" "-Zunstable-options"
stdout: none
--- stderr -------------------------------
error[E0570]: `"vectorcall"` is not a supported ABI for the current target
##[error]  --> /checkout/tests/rustdoc-json/fns/abi.rs:19:12
   |
19 | pub extern "vectorcall" fn abi_vectorcall() {}
   |            ^^^^^^^^^^^^

error[E0570]: `"vectorcall-unwind"` is not a supported ABI for the current target
##[error]  --> /checkout/tests/rustdoc-json/fns/abi.rs:22:12
   |
22 | pub extern "vectorcall-unwind" fn abi_vectorcall_unwind() {}
   |            ^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0570`.
------------------------------------------


---- [rustdoc-json] tests/rustdoc-json/methods/abi.rs stdout ----

error: rustdoc failed!
status: exit status: 1
command: cd "/checkout/obj/build/aarch64-unknown-linux-gnu/test/rustdoc-json/methods/abi" && env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustdoc" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/rustdoc-json/methods/abi/auxiliary" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/rustdoc-json/methods/abi" "--deny" "warnings" "/checkout/tests/rustdoc-json/methods/abi.rs" "-A" "internal_features" "--output-format" "json" "-Zunstable-options"
stdout: none
--- stderr -------------------------------
error[E0570]: `"vectorcall"` is not a supported ABI for the current target
##[error]  --> /checkout/tests/rustdoc-json/methods/abi.rs:23:16
   |
23 |     pub extern "vectorcall" fn abi_vectorcall() {}
   |                ^^^^^^^^^^^^

error[E0570]: `"vectorcall-unwind"` is not a supported ABI for the current target
##[error]  --> /checkout/tests/rustdoc-json/methods/abi.rs:26:16
   |
26 |     pub extern "vectorcall-unwind" fn abi_vectorcall_unwind() {}
   |                ^^^^^^^^^^^^^^^^^^^

error[E0570]: `"vectorcall"` is not a supported ABI for the current target
##[error]  --> /checkout/tests/rustdoc-json/methods/abi.rs:46:12
   |
46 |     extern "vectorcall" fn trait_abi_vectorcall() {}
   |            ^^^^^^^^^^^^

error[E0570]: `"vectorcall-unwind"` is not a supported ABI for the current target
##[error]  --> /checkout/tests/rustdoc-json/methods/abi.rs:49:12
   |
49 |     extern "vectorcall-unwind" fn trait_abi_vectorcall_unwind() {}
   |            ^^^^^^^^^^^^^^^^^^^

```

</details>

workingjubilee commented:
@aDotInTheVoid Mm. I... am just gonna remove the `extern "vectorcall"` instances from these tests if that's alright? I don't see what useful coverage you're getting by testing for them?

aDotInTheVoid commented:
> I don't see what useful coverage you're getting by testing for them if you already have e.g. `extern "system"` covered?

It's checking that we produces the `rustdoc_json_types::Abi::Other` variant for ABI's that don't have a dedicated variant in rustdoc-json.

https://github.com/rust-lang/rust/blob/91fad92585b2dafc52a074e502b2a6c1f093ca35/src/librustdoc/json/conversions.rs#L365-L378

https://github.com/rust-lang/rust/blob/91fad92585b2dafc52a074e502b2a6c1f093ca35/src/rustdoc-json-types/lib.rs#L771-L794

Could you instead change the tests to use some other unstable ABI here? But it's fine to only test this once (with both the `-unwind` and not cases), the current thing of testing every kind of abi everywhere an abi can appear is kind of excessive and provides very little value.

workingjubilee commented:
@aDotInTheVoid Well. Most other unstable ABIs *don't have* "unwind" cases.

workingjubilee committed:
rustdoc-json: Rearrange deck chairs in ABI testing

We move the vectorcall ABI tests into their own file which is now
only run on x86-64, while replacing them with rust-cold ABI tests
so that aarch64 hosts continue to test an unstable ABI.

A better solution might be cross-compiling or something but
I really don't have time for that right now.

workingjubilee commented:
I have pushed an alternative which does some rearrangement so that we don't have complications due to aarch64 hosts.

Nadrieril commented:
I don't know enough about ABI to review this

r? compiler
